src/Cli/dotnet/Installer/Windows/MsiPackageCache.cs

```diff
@@ -0,0 +1,240 @@
+            s_DefaultInheritance, PropagationFlags.None, AccessControlType.Allow);
+
+        /// <summary>
+        /// ACL rule associated with the Everyone SID.
+        /// </summary>
+        private static readonly FileSystemAccessRule s_EveryoneRule = new FileSystemAccessRule(s_EveryoneSid, FileSystemRights.ReadAndExecute,
+            s_DefaultInheritance, PropagationFlags.None, AccessControlType.Allow);
+
+        /// <summary>
+        /// ACL rule associated with the Local SYSTEM SID.
+        /// </summary>
+        private static readonly FileSystemAccessRule s_LocalSystemRule = new FileSystemAccessRule(s_LocalSystemSid, FileSystemRights.FullControl,
+            s_DefaultInheritance, PropagationFlags.None, AccessControlType.Allow);
+
+        /// <summary>
+        /// ACL rule associated with the built-in users SID.
+        /// </summary>
+        private static readonly FileSystemAccessRule s_UsersRule = new FileSystemAccessRule(s_UsersSid, FileSystemRights.ReadAndExecute,
+            s_DefaultInheritance, PropagationFlags.None, AccessControlType.Allow);
+
+        /// <summary>
+        /// The root directory of the package cache where MSI workload packs are stored.
+        /// </summary>
+        public readonly string PackageCacheRoot;
+
+        public MsiPackageCache(InstallElevationContextBase elevationContext, ISetupLogger logger,
+            string packageCacheRoot = null) : base(elevationContext, logger)
+        {
+            PackageCacheRoot = string.IsNullOrWhiteSpace(packageCacheRoot)
+                ? Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.CommonApplicationData), "dotnet", "workloads")
+                : packageCacheRoot;
+        }
+
+        /// <summary>
+        /// Creates the root package cache directory if it does not exist and configures the directory's ACLs. ACLs are still configured
+        /// if the directory exists.
+        /// </summary>
+        private void CreateRootDirectory()
+        {
+            CreateSecureDirectory(PackageCacheRoot);
+        }
+
+        /// <summary>
+        /// Creates the specified directory and secures it by configuring access rules (ACLs). If the parent
+        /// of the directory does not exist, it recursively walks the path back to ensure each parent directory
+        /// is created with the proper ACLs and inheritance settings.
+        /// </summary>
+        /// <param name="path">The path of the directory to create.</param>
+        private void CreateSecureDirectory(string path)
+        {
+            if (!Directory.Exists(path))
+            {
+                CreateSecureDirectory(Directory.GetParent(path).FullName);
+
+                DirectorySecurity directorySecurity = new();
+                directorySecurity.SetAccessRule(s_AdministratorRule);
+                directorySecurity.SetGroup(s_LocalSystemSid);
+                directorySecurity.CreateDirectory(path);
+                SecureDirectory(path);
+            }
```

Comment:
If the dir already exists we assume that it's secure without checking any access rules? Is that okay?